ffc/ir/uflacs/analysis/factorization.py

```diff
@@ -299,39 +299,45 @@ def compute_argument_factorization(S, rank):
 
     assert len(F.nodes) == len(F.e2i)
 
-    # Find the (only) node in S that is marked as 'target'
-    # Should always be the last one.
+    # Prepare a mapping from component of expression to factors
+    factors = {}
     S_targets = [i for i, v in S.nodes.items() if v.get('target', False)]
-    assert len(S_targets) == 1
-    S_target = S_targets[0]
-
-    # Get the factorizations of the target values
-    if S.nodes[S_target]['factors'] == {}:
-        if rank == 0:
-            # Functionals and expressions: store as no args * factor
-            factors = {(): F.e2i[S.nodes[S_target]['expression']]}
+
+    for S_target in S_targets:
+        # Get the factorizations of the target values
+        if S.nodes[S_target]['factors'] == {}:
+            if rank == 0:
+                # Functionals and expressions: store as no args * factor
+                for comp in S.nodes[S_target]["component"]:
+                    factors[comp] = {(): F.e2i[S.nodes[S_target]['expression']]}
+            else:
+                # Zero form of arity 1 or higher: make factors empty
+                pass
         else:
-            # Zero form of arity 1 or higher: make factors empty
-            factors = {}
-    else:
-        # Forms of arity 1 or higher:
-        # Map argkeys from indices into SV to indices into AV,
-        # and resort keys for canonical representation
-        factors = {
-            tuple(sorted(arg_indices.index(si) for si in argkey)): fi
-            for argkey, fi in S.nodes[S_target]['factors'].items()
-        }
-    # Expecting all term keys to have length == rank
-    # (this assumption will eventually have to change if we
-    # implement joint bilinear+linear form factorization here)
-    assert all(len(k) == rank for k in factors)
+            # Forms of arity 1 or higher:
+            # Map argkeys from indices into SV to indices into AV,
+            # and resort keys for canonical representation
+            for argkey, fi in S.nodes[S_target]['factors'].items():
```

Comment:
If find `fi` a bit difficult to parse. Is there are better name?

Reply:
`fi` is used on many places. There is a comment explaining the meaning. It is an index (in factorised graph) of current argument factor.